File: 0054-spiral-matrix/0054-spiral-matrix.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        if (matrix.length == 0 || matrix[0].length == 0) return res;
        
        int top = 0, bottom = matrix.length - 1;
        int left = 0, right = matrix[0].length - 1;

        while (left <= right && top <= bottom) {
            // Traverse from left to right
            for (int col = left; col <= right; col++) {
                res.add(matrix[top][col]);
            }
            top++;

            // Traverse from top to bottom
            for (int row = top; row <= bottom; row++) {
                res.add(matrix[row][right]);
            }
            right--;

            if (top <= bottom) {
                // Traverse from right to left
                for (int col = right; col >= left; col--) {
                    res.add(matrix[bottom][col]);
                }
                bottom--;
            }

            if (left <= right) {
                // Traverse from bottom to top
                for (int row = bottom; row >= top; row--) {
                    res.add(matrix[row][left]);
                }
                left++;
            }
        }

        return res;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rows = len(matrix)
        cols = len(matrix[0])
        res = []

        top, left = 0,0
        bottom, right = rows-1,cols-1

        # while left <= right and top <= bottom:
        while len(res) != rows * cols:
            for col in range(left,right+1):
                res.append(matrix[top][col])
            top += 1

            for row in range(top,bottom+1):
                res.append(matrix[row][right])
            right -= 1

            if top <= bottom:
                for col in range(right,left-1,-1):
                    res.append(matrix[bottom][col])
                bottom -= 1
            
            if left <= right:
                for row in range(bottom,top-1,-1):
                    res.append(matrix[row][left])
                left += 1
        return res